server/src/main/java/org/opensearch/index/store/remote/directory/RemoteSearchDirectory.java

```diff
@@ -0,0 +1,132 @@
+import org.apache.lucene.store.Lock;
+import org.apache.lucene.store.NoLockFactory;
+import org.opensearch.index.store.RemoteSegmentStoreDirectory;
+import org.opensearch.index.store.remote.file.OnDemandBlockSearchIndexInput;
+import org.opensearch.index.store.remote.utils.TransferManager;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * a Directory implementation that can read directly from index segments metadata stored in the remote store.
+ *
+ * @opensearch.internal
+ */
+public final class RemoteSearchDirectory extends Directory {
+    private final FSDirectory localStoreDir;
+    private final TransferManager transferManager;
+
+    private final Map<String, RemoteSegmentStoreDirectory.UploadedSegmentMetadata> uploadedSegmentMetadataMap;
+
+    public RemoteSearchDirectory(Map<String, RemoteSegmentStoreDirectory.UploadedSegmentMetadata> uploadedSegmentMetadataMap,
+                                 FSDirectory localStoreDir, TransferManager transferManager) {
+        this.localStoreDir = localStoreDir;
+        this.transferManager = transferManager;
+        this.uploadedSegmentMetadataMap = uploadedSegmentMetadataMap;
+    }
+
+    @Override
+    public String[] listAll() throws IOException {
+        return uploadedSegmentMetadataMap.keySet().toArray(new String[0]);
+    }
+
+    @Override
+    public void deleteFile(String name) throws IOException {}
+
+    @Override
+    public long fileLength(String name) throws IOException {
+        return uploadedSegmentMetadataMap.get(name).getLength();
+    }
+
+    @Override
+    public IndexOutput createOutput(String name, IOContext context) {
+        return NoopIndexOutput.INSTANCE;
+    }
+
+    @Override
+    public IndexInput openInput(String name, IOContext context) throws IOException {
+        final RemoteSegmentStoreDirectory.UploadedSegmentMetadata uploadedSegmentMetadata = uploadedSegmentMetadataMap.get(name);
+        return new OnDemandBlockSearchIndexInput(uploadedSegmentMetadata, localStoreDir, transferManager);
+    }
+
+    @Override
+    public void close() throws IOException {
+        localStoreDir.close();
+    }
+
+    @Override
+    public Set<String> getPendingDeletions() throws IOException {
+        return Collections.emptySet();
+    }
+
+    @Override
+    public IndexOutput createTempOutput(String prefix, String suffix, IOContext context) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public void sync(Collection<String> names) throws IOException {}
+
+    @Override
+    public void syncMetaData() {}
+
+    @Override
+    public void rename(String source, String dest) throws IOException {}
+
+    @Override
+    public Lock obtainLock(String name) throws IOException {
+        return NoLockFactory.INSTANCE.obtainLock(null, null);
+    }
+
+    static class NoopIndexOutput extends IndexOutput {
```

Comment:
Let's pull the common functionalities including this class out to a new class `RemoteDirectory` which can be extended by `RemoteSearchDirectory` and `RemoteSnapshotDirectory`. 

```
public abstract class RemoteDirectory extends Directory {
    ...
    <All empty and Noop based methods>
    ...
    ...
    static class NoopIndexOutput extends IndexOutput {
   }
}
```

```
public final class RemoteSnapshotDirectory extends RemoteDirectory {}
public final class RemoteSearchDirectory extends RemoteDirectory {}

```
That should prevent a bunch of NoOp duplication across classes, and can be overridden by specific implementations in future phases, if needed.

Reply:
RemoteDirectory is another class present [ref](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/store/RemoteDirectory.java), but I see your point we can take out the common part to another class.